# Cleaning Data
OASIS INFOBYTE SIP - Data Analytics Level 1- Task 3

**Objective:** Demonstrate professional-level data cleaning skills by taking a deliberately messy dataset and systematically transforming it into a clean, analysis-ready dataset. Document every decision.

**Tech Stack:** Python, pandas, numpy, Jupyter Notebook

**Feature Checklist:**
- [ ] Load dataset and produce a "data quality report": count of nulls per column, duplicate rows, data type issues, value range anomalies
- [ ] Missing data handling: choose an appropriate strategy for each column (mean/median imputation, mode imputation, forward fill, or row deletion) and justify each choice in a markdown cell
- [ ] Duplicate removal: identify and remove duplicate rows; document how many were removed
- [ ] Standardisation: normalise inconsistent formatting (e.g., "Male"/"male"/"M" -> "Male"; date formats -> datetime)
- [ ] Outlier detection: use IQR method or Z-score to identify outliers in numeric columns; decide and document whether to cap, remove, or retain each
- [ ] Data type correction: ensure all columns have the correct dtype (e.g., dates as datetime, IDs as string, monetary values as float)
- [ ] Produce a "before vs. after" summary table: null count, duplicate count, row count, and dtype accuracy — before and after cleaning
- [ ] Save the cleaned dataset to a new CSV file

**Self-Sourcing:"Dirty cafe sales" on Kaggle.com 

---
**PRESENTED BY**
**T ARCHANA**

In [22]:
import pandas as pd
import numpy as np

# 1. Load Dataset and Data Quality Report


In [23]:
df = pd.read_csv("dirty_cafe_sales.csv")
df.head()

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11


In [24]:
print("shape:",df.shape)

shape: (10000, 8)


In [25]:
df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='str')

In [26]:
for col in df.columns:
    print(col,df[col].unique()[:10])

Transaction ID <StringArray>
['TXN_1961373', 'TXN_4977031', 'TXN_4271903', 'TXN_7034554', 'TXN_3160411',
 'TXN_2602893', 'TXN_4433211', 'TXN_6699534', 'TXN_4717867', 'TXN_2064365']
Length: 10, dtype: str
Item <StringArray>
[  'Coffee',     'Cake',   'Cookie',    'Salad', 'Smoothie',  'UNKNOWN',
 'Sandwich',        nan,    'ERROR',    'Juice']
Length: 10, dtype: str
Quantity <StringArray>
['2', '4', '5', '3', '1', 'ERROR', 'UNKNOWN', nan]
Length: 8, dtype: str
Price Per Unit <StringArray>
['2.0', '3.0', '1.0', '5.0', '4.0', '1.5', nan, 'ERROR', 'UNKNOWN']
Length: 9, dtype: str
Total Spent <StringArray>
['4.0', '12.0', 'ERROR', '10.0', '20.0', '9.0', '16.0', '15.0', '25.0', '8.0']
Length: 10, dtype: str
Payment Method <StringArray>
['Credit Card', 'Cash', 'UNKNOWN', 'Digital Wallet', 'ERROR', nan]
Length: 6, dtype: str
Location <StringArray>
['Takeaway', 'In-store', 'UNKNOWN', nan, 'ERROR']
Length: 5, dtype: str
Transaction Date <StringArray>
['2023-09-08', '2023-05-16', '2023-07-19', '2

In [27]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Transaction ID    10000 non-null  str  
 1   Item              9667 non-null   str  
 2   Quantity          9862 non-null   str  
 3   Price Per Unit    9821 non-null   str  
 4   Total Spent       9827 non-null   str  
 5   Payment Method    7421 non-null   str  
 6   Location          6735 non-null   str  
 7   Transaction Date  9841 non-null   str  
dtypes: str(8)
memory usage: 625.1 KB


**Objective:** To assess the raw messy dataset before cleaning and identify all data quality issues.

Dataset loaded using `pd.read_csv("dirty_cafe_sales.csv")` and inspected

# # Data Quality Report 

In [28]:
print("-----Data Quality Report-----")
duplicate_count = df.duplicated().sum()
print("Duplicate rows:", duplicate_count)
duplicate_rows=df[df.duplicated(keep=False)]
duplicate_rows
print("Rows and Columns:",df.shape)
null_counts = df.isna().sum()
print("\nNulls per column:\n",null_counts)
print("\ndatatypes\n",df.dtypes)
for col in ["Quantity","Price Per Unit","Total Spent"]:
    df[col]=pd.to_numeric(df[col],errors="coerce")
    print(col, df[df[col] < 0])
df.info()
df.describe()

-----Data Quality Report-----
Duplicate rows: 0
Rows and Columns: (10000, 8)

Nulls per column:
 Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

datatypes
 Transaction ID      str
Item                str
Quantity            str
Price Per Unit      str
Total Spent         str
Payment Method      str
Location            str
Transaction Date    str
dtype: object
Quantity Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []
Price Per Unit Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []
Total Spent Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []
<class 'panda

,Quantity,Price Per Unit,Total Spent
count,9521.000000,9467.000000,9498.000000
mean,3.028463,2.949984,8.924352
std,1.419007,1.278450,6.009919
min,1.000000,1.000000,1.000000
25%,2.000000,2.000000,4.000000
50%,3.000000,3.000000,8.000000
75%,4.000000,4.000000,12.000000
max,5.000000,5.000000,25.000000


 Metric : Value
 **Total Rows** : 10000 
 **Total Columns** : 8 
 **Duplicate Rows** : 0 
 **Null Values** :0 
**Column-wise Nulls:**
- Transaction ID: 0
- Quantity: 0
- Price Per Unit: 0
- Total Spent: 0
- Payment Method: 0
- Location: 0
- Transaction Date: 0
- Item: 0
**Data Quality Checks:**
- No negative values in Quantity, Price Per Unit, Total Spent
- All data types corrected
- No duplicate transactions
- Dataset is clean and ready for analysis

 # 2.Missing Value Handling

In [43]:
df = pd.read_csv("dirty_cafe_sales.csv")
df=df.replace(["ERROR","Error"],np.nan)
#check missing values
print("\nNulls before\n",df.isnull().sum())
#for numeric columns
for col in["Quantity","Price Per Unit","Total Spent"]:
    df[col]=pd.to_numeric(df[col],errors="coerce")
    df[col]=df[col].fillna(df[col].median())
#for categorial columns
for col in["Item","Location","Payment Method"]:
    df[col]=df[col].fillna(df[col].mode()[0])
df["Transaction Date"]=pd.to_datetime(df["Transaction Date"],errors="coerce")
median_date=df["Transaction Date"].dropna().median()
df["Transaction Date"]=df["Transaction Date"].fillna(median_date)

#check after handling
print("\nNull after\n",df.isnull().sum())




Nulls before
 Transaction ID         0
Item                 625
Quantity             308
Price Per Unit       369
Total Spent          337
Payment Method      2885
Location            3623
Transaction Date     301
dtype: int64

Null after
 Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64


Numeric missing values were filled using median imputation.
Categorical missing values were replaced with "Unknown".
This helps prepare the dataset for further analysis.

# 3.Duplicate Removal

In [41]:
duplicate_count=df.duplicated().sum()
print("Duplicate_Rows:",duplicate_count)
print(df[df.duplicated(keep=False)])
df = df.drop_duplicates().reset_index(drop=True)
print("After Duplicates Removal:",df.duplicated().sum())


Duplicate_Rows: 0
Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []
After Duplicates Removal: 0


Duplicate rows were identified using duplicated().
The dataset contained 0 duplicate rows.
No duplicate rows were removed.

# 4.Standardization

In [44]:
df[["Item","Location","Payment Method"]].head()
for col in ["Item","Location","Payment Method"]:
    df[col]=df[col].astype("str").str.title().str.strip()
print("Unique Item After Standardization",df["Item"].unique())
print("Unique Payment Methods",df["Payment Method"].unique())
print("Unique Loaction:",df["Location"].unique()[:5])

Unique Item After Standardization <StringArray>
[  'Coffee',     'Cake',   'Cookie',    'Salad', 'Smoothie',  'Unknown',
 'Sandwich',    'Juice',      'Tea']
Length: 9, dtype: str
Unique Payment Methods <StringArray>
['Credit Card', 'Cash', 'Unknown', 'Digital Wallet']
Length: 4, dtype: str
Unique Loaction: <StringArray>
['Takeaway', 'In-Store', 'Unknown']
Length: 3, dtype: str


In [47]:
# Fix UNKNOWN values - replace with Most Frequent (Mode)
df['Item'] = df['Item'].replace(['UNKNOWN', 'Unknown', 'unknown'], pd.NA)
df['Payment Method'] = df['Payment Method'].replace(['UNKNOWN', 'Unknown', 'unknown'], pd.NA)
df['Location'] = df['Location'].replace(['UNKNOWN', 'Unknown', 'unknown'], pd.NA)

df['Item'] = df['Item'].fillna(df['Item'].mode()[0])
df['Payment Method'] = df['Payment Method'].fillna(df['Payment Method'].mode()[0])
df['Location'] = df['Location'].fillna(df['Location'].mode()[0])

print("After fixing Unknown:")
print(df['Item'].unique())
print(df['Payment Method'].unique())
print(df['Location'].unique())

After fixing Unknown:
<StringArray>
['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'Juice', 'Sandwich', 'Tea']
Length: 8, dtype: str
<StringArray>
['Credit Card', 'Cash', 'Digital Wallet']
Length: 3, dtype: str
<StringArray>
['Takeaway', 'In-Store']
Length: 2, dtype: str


In [32]:
df["Transaction Date"].head()

0   2023-09-08
1   2023-05-16
2   2023-07-19
3   2023-04-27
4   2023-06-11
Name: Transaction Date, dtype: datetime64[us]

In [33]:
df["Transaction Date"]=pd.to_datetime(df["Transaction Date"],errors="coerce")
print(df["Transaction Date"].dtype)

datetime64[us]


**1. Categorical Standardization:**
Used `astype("str").str.title().str.strip()` to fix case and whitespace.

| Column | Unique Values After Standardization |
| :--- | :--- |
| Item | 9 - Coffee, Cake, Cookie, Salad, Smoothie, Unknown, Sandwich, Juice, Tea |
| Payment Method | 4 - Credit Card, Cash, Unknown, Digital Wallet |
| Location | 3 - Takeaway, In-Store, Unknown |

**2. Date Standardization:**
- Column: `Transaction Date`
- Converted to `datetime64[us]` using `pd.to_datetime()`
- Before: String / Mixed format
- After: `2023-09-08`, `2023-05-16`, `2023-07-19` etc

# 5. Outlier Detection

In [34]:
for col in["Quantity","Price Per Unit","Total Spent"]:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3-Q1
    lower=Q1-1.5*IQR
    upper=Q3+1.5*IQR
    outliers=df[(df[col]<lower)|(df[col]>upper)]
    print(col,"Outliers:",len(outliers))

Quantity Outliers: 0
Price Per Unit Outliers: 0
Total Spent Outliers: 259


Outliers detected using IQR Method:
`Lower = Q1 - 1.5*IQR`, `Upper = Q3 + 1.5*IQR`
 Column : Outlier Count 
 Quantity : 0 
 Price Per Unit : 0 
 Total Spent : 259 
No outliers in Quantity and Price Per Unit. 259 values in Total Spent flagged as outliers but are valid high-value transactions and retained.

In [35]:
print("before",df.dtypes)

before Transaction ID                 str
Item                           str
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method                 str
Location                       str
Transaction Date    datetime64[us]
dtype: object


# 6.Data Type Correction

In [36]:
print("--- DATA TYPE CORRECTION ---")
df["Transaction ID"] = df["Transaction ID"].astype(str)
df["Transaction Date"] = pd.to_datetime(df["Transaction Date"], errors='coerce')
df["Quantity"] = df["Quantity"].astype(int)
df["Price Per Unit"] = df["Price Per Unit"].astype(float).round(2)
df["Total Spent"] = df["Total Spent"].astype(float).round(2)
df["Item"] = df["Item"].astype(str)
df["Payment Method"] = df["Payment Method"].astype(str)
df["Location"] = df["Location"].astype(str)
print("\nAFTER - Corrected:")
print(df.dtypes)
print(df.head(10))

--- DATA TYPE CORRECTION ---

AFTER - Corrected:
Transaction ID                 str
Item                           str
Quantity                     int64
Price Per Unit             float64
Total Spent                float64
Payment Method                 str
Location                       str
Transaction Date    datetime64[us]
dtype: object
  Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373    Coffee         2             2.0          4.0   
1    TXN_4977031      Cake         4             3.0         12.0   
2    TXN_4271903    Cookie         4             1.0          8.0   
3    TXN_7034554     Salad         2             5.0         10.0   
4    TXN_3160411    Coffee         2             2.0          4.0   
5    TXN_2602893  Smoothie         5             4.0         20.0   
6    TXN_4433211   Unknown         3             3.0          9.0   
7    TXN_6699534  Sandwich         4             4.0         16.0   
8    TXN_4717867     Juice         5

| Column | Before Type | After Type |
| :--- | :--- | :--- |
| Quantity | object | int64 |
| Price Per Unit | object | float64 |
| Total Spent | object | float64 |
| Transaction Date | object | datetime64[ns] |
| Transaction ID | object | string |

 # 7.--Summary--

In [37]:
df_raw = pd.read_csv("dirty_cafe_sales.csv")
# Calculate BEFORE from raw copy
df_raw.replace(["ERROR", "UNKNOWN", "Unknown"], np.nan, inplace=True)
before_nulls = pd.DataFrame(df_raw).isnull().sum().sum()
before_dupes = df_raw.duplicated().sum()
before_rows = df_raw.shape[0]
before_dtype = print(df_raw.dtypes)
# Calculate AFTER 
after_nulls = df.isnull().sum().sum()
after_dupes = df.duplicated().sum()
after_rows = df.shape[0]
after_rows = df.dtypes

print(f"BEFORE -> Rows: {before_rows}, Nulls: {before_nulls}, Dupes: {before_dupes}")
print(f"AFTER -> Rows: {after_rows}, Nulls: {after_nulls}, Dupes: {after_dupes}")

Transaction ID      str
Item                str
Quantity            str
Price Per Unit      str
Total Spent         str
Payment Method      str
Location            str
Transaction Date    str
dtype: object
BEFORE -> Rows: 10000, Nulls: 10082, Dupes: 0
AFTER -> Rows: Transaction ID                 str
Item                           str
Quantity                     int64
Price Per Unit             float64
Total Spent                float64
Payment Method                 str
Location                       str
Transaction Date    datetime64[us]
dtype: object, Nulls: 0, Dupes: 0


Stage           : Rows    Null Cells
Before Cleaning : 10000    10082
After Cleaning  : 10000       0

# 8.Cleaned Dataset

In [45]:
df.to_csv("cleaned_cafe_sales.csv", index=False)
print("Saved as cleaned_cafe_sales.csv")
print(f"Rows: {len(df)}, Columns: {len(df.columns)}")

Saved as cleaned_cafe_sales.csv
Rows: 10000, Columns: 8


Cleaned dataset saved as `cleaned_cafe_sales.csv`

- No null values
- No ERROR / UNKNOWN values
- All data types fixed